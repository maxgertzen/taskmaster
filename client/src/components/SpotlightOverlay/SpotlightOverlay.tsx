import { FC, useEffect, useState } from 'react';

import { useSpotlightStore } from '../../store/store';

import { StyledSpotlightOverlay } from './SpotlightOverlay.styled';

export const SpotlightOverlay: FC = () => {
  const target = useSpotlightStore((state) => state.target);
  const [targetRect, setTargetRect] = useState<DOMRect | null>(null);

  useEffect(() => {
    if (target) {
      const targetElement = document.querySelector(
        `[data-spotlight="${target}"]`
      );
      setTargetRect(targetElement?.getBoundingClientRect() ?? null);
      return;
    }
    setTargetRect(null);
  }, [target]);

  return <StyledSpotlightOverlay targetRect={targetRect} />;
};
