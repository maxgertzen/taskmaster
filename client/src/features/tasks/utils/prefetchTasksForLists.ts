import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
