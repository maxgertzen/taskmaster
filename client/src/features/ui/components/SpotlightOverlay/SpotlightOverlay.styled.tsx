import styled from '@emotion/styled';

export const StyledSpotlightOverlay = styled('div')<{
  targetRect: DOMRect | null;
}>(({ targetRect }) => ({
  position: 'fixed',
  top: 0,
  left: 0,
  width: '100vw',
  height: '100vh',
  background: 'rgba(0, 0, 0, 0.5)',
  pointerEvents: 'auto',
  zIndex: targetRect ? 9999 : -1,
  opacity: targetRect ? 1 : 0,
  backdropFilter: 'blur(5px)',
  maskImage: targetRect
    ? `radial-gradient(circle at ${targetRect.left + targetRect.width / 2}px ${
        targetRect.top + targetRect.height / 2
      }px, transparent ${Math.max(targetRect.width, targetRect.height)}px, black ${Math.max(
        targetRect.width,
        targetRect.height
      )}px)`
    : 'none',
  WebkitMaskImage: targetRect
    ? `radial-gradient(circle at ${targetRect.left + targetRect.width / 2}px ${
        targetRect.top + targetRect.height / 2
      }px, transparent ${Math.max(targetRect.width, targetRect.height)}px, black ${Math.max(
        targetRect.width,
        targetRect.height
      )}px)`
    : 'none',
  transition: `mask-image 0.3s ease-in, -webkit-mask-image 0.3s ease-in`,
}));
