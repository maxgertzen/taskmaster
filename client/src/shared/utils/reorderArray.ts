export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
