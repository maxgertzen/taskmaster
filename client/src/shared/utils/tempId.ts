export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
