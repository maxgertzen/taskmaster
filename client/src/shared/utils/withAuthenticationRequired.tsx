import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
