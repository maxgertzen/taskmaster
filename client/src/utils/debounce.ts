export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
