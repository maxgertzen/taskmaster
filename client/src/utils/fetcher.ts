export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
