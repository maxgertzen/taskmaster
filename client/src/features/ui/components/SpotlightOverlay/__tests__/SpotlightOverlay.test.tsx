import { describe, it, expect, vi, beforeEach } from 'vitest';

import {
  mockSpotlightStore,
  mockUseSpotlightStore,
} from '@/tests/__mocks__/storeMocks';
import { render } from '@/tests/utils';

import { SpotlightOverlay } from '../SpotlightOverlay';

vi.mock('@/shared/store/spotlightStore', () => ({
  useSpotlightStore: mockSpotlightStore,
}));

describe('SpotlightOverlay', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders without target', () => {
    mockUseSpotlightStore.mockReturnValue(null);
    render(<SpotlightOverlay />);
  });

  it('should match snapshot', () => {
    mockUseSpotlightStore.mockReturnValue(null);
    const { container } = render(<SpotlightOverlay />);
    expect(container).toMatchSnapshot();
  });
});
