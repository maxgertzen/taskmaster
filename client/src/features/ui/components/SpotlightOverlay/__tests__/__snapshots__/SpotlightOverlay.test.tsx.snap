// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpotlightOverlay > handles non-existent target element 1`] = `
<div>
  <div
    class="css-12ictld"
  />
</div>
`;

exports[`SpotlightOverlay > renders without target 1`] = `
<div>
  <div
    class="css-12ictld"
  />
</div>
`;

exports[`SpotlightOverlay > should match snapshot 1`] = `
<div>
  <div
    class="css-12ictld"
  />
</div>
`;
